### PR TITLE
Refactor & fix some minor bugs in alias elimination & add consistency check

### LINKED ISSUE #812: Add check_consistency
`check_consistency(sys, verbose=true)` should check the consistency of a system, and report some basic statistics when `verbose=true`.

### LINKED ISSUE #933: `structural_simplify` throws for a high index circuit model
This is a full model version of a problem described in #931 
```julia
using ModelingToolkit, DifferentialEquations

function Pin(;name)
    @variables v(t) i(t)
    ODESystem(Equation[], t, [v, i], [], name=name, defaults=[v=>1.0, i=>1.0])
end
  
function Ground(;name)
    @named g = Pin()
    eqs = [g.v ~ 0]
    ODESystem(eqs, t, [], [], systems=[g], name=name)
end

function Resistor(;name, R = 1.0)
    val = R
    @named p = Pin()
    @named n = Pin()
    @variables v(t)
    @parameters R
    eqs = [
           v ~ p.v - n.v
           0 ~ p.i + n.i
           v ~ p.i * R
          ]
    ODESystem(eqs, t, [v], [R], systems=[p, n], defaults=Dict(R => val), name=name)
end

function Inductor(; name, L = 1.0)
    val = L
    @named p = Pin()
    @named n = Pin()
    @variables v(t) i(t)
    @parameters L
    D = Differential(t)
    eqs = [
           v ~ p.v - n.v
           0 ~ p.i + n.i
           i ~ p.i
           D(i) ~ v / L
          ]
    ODESystem(eqs, t, [v, i], [L], systems=[p, n], defaults=Dict(L => val), name=name)
end

function ConstantVoltage(;name, V = 1.0)
    val = V
    @named p = Pin()
    @named n = Pin()
    @parameters V
    eqs = [
           V ~ p.v - n.v
           0 ~ p.i + n.i
          ]
    ODESystem(eqs, t, [], [V], systems=[p, n], defaults=Dict(V => val), name=name)
end

function connect_pins(ps...)
    eqs = [
           0 ~ sum(p->p.i, ps) # KCL
          ]
    # KVL
    for i in 1:length(ps)-1
        push!(eqs, ps[i].v ~ ps[i+1].v)
    end

    return eqs
end

@parameters t

@named V1 = ConstantVoltage(V=10.0)
@named R1 = Resistor(R=1.0)
@named L1 = Inductor(L=8.0e-9)
@named L2 = Inductor(L=2.0e-9)
@named GND = Ground()

ckt = [
       connect_pins(V1.p, R1.p)
       connect_pins(R1.n, L1.p)
       connect_pins(L1.n, L2.p)
       connect_pins(V1.n, L2.n, GND.g)
      ]

@named ckt_model = ODESystem(ckt, t, systems=[V1, R1, L1, L2, GND])

sys = structural_simplify(ckt_model)
```
`structural_simplify` fails with this error:
```text
ERROR: LoadError: 

... Internal error in Tearing.jl: vs[1] = 0.

Stacktrace:
  [1] error(::String, ::Int64, ::String, ::Int64, ::String)
    @ Base ./error.jl:42
  [2] initAlgebraicSystem(td::ModelingToolkit.StructuralTransformations.TraverseDAG, es::Vector{Int64}, vs::Vector{Int64}, eSolvedFixed::Vector{Int64}, vSolvedFixed::Vector{Int64}, vTearFixed::Vector{Int64})
    @ ModelingToolkit.StructuralTransformations ~/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/bipartite_tearing/modia_tearing.jl:116
  [3] tearEquations!(td::ModelingToolkit.StructuralTransformations.TraverseDAG, Gsolvable::Vector{Vector{Int64}}, es::Vector{Int64}, vs::Vector{Int64}; eSolvedFixed::Vector{Int64}, vSolvedFixed::Vector{Int64}, vTearFixed::Vector{Int64})
    @ ModelingToolkit.StructuralTransformations ~/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/bipartite_tearing/modia_tearing.jl:320
  [4] tearEquations!
    @ ~/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/bipartite_tearing/modia_tearing.jl:320 [inlined]
  [5] (::ModelingToolkit.StructuralTransformations.var"#28#30"{Vector{Int64}, Vector{Int64}, BipartiteGraph{Int64, Nothing}, BipartiteGraph{Int64, Nothing}, SystemStructure})(c::Vector{Int64})
    @ ModelingToolkit.StructuralTransformations ~/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/tearing.jl:16
  [6] iterate
    @ ./generator.jl:47 [inlined]
  [7] collect_to!
    @ ./array.jl:724 [inlined]
  [8] collect_to_with_first!(dest::Vector{NTuple{4, Vector{Int64}}}, v1::NTuple{4, Vector{Int64}}, itr::Base.Generator{Vector{Vector{Int64}}, ModelingToolkit.StructuralTransformations.var"#28#30"{Vector{Int64}, Vector{Int64}, BipartiteGraph{Int64, Nothing}, BipartiteGraph{Int64, Nothing}, SystemStructure}}, st::Int64)
    @ Base ./array.jl:702
  [9] _collect(c::Vector{Vector{Int64}}, itr::Base.Generator{Vector{Vector{Int64}}, ModelingToolkit.StructuralTransformations.var"#28#30"{Vector{Int64}, Vector{Int64}, BipartiteGraph{Int64, Nothing}, BipartiteGraph{Int64, Nothing}, SystemStructure}}, #unused#::Base.EltypeUnknown, isz::Base.HasShape{1})
    @ Base ./array.jl:696
 [10] collect_similar
    @ ./array.jl:606 [inlined]
 [11] map
    @ ./abstractarray.jl:2294 [inlined]
 [12] tear_graph(sys::ODESystem)
    @ ModelingToolkit.StructuralTransformations ~/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/tearing.jl:11
 [13] tearing(sys::ODESystem; simplify::Bool)
    @ ModelingToolkit.StructuralTransformations ~/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/tearing.jl:221
 [14] tearing(sys::ODESystem)
    @ ModelingToolkit.StructuralTransformations ~/.julia/packages/ModelingToolkit/7KM6m/src/structural_transformation/tearing.jl:221
 [15] structural_simplify(sys::ODESystem)
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/7KM6m/src/systems/abstractsystem.jl:526
 [16] top-level scope
    @ ~/test.jl:86
in expression starting at /Users/szymon/test.jl:86
```
Julia 1.6.0 with `Project.toml`
```toml
[deps]
DifferentialEquations = "0c46a032-eb83-5123-abaf-570d42b7fbaa"
ModelingToolkit = "961ee093-0014-501f-94e3-6117800e7a78"
```

### PR DESCRIPTION
**Pull request by YingboMa**
Fix #933
Close #812

### CODE REVIEW

File: src/bipartite_graph.jl [hunk 1]
```diff
@@ -66,13 +66,13 @@ badjlist = [[1,2,5,6],[3,4,6]]
...
+mutable struct BipartiteGraph{I<:Integer,F<:Vector{Vector{I}},B<:Union{Vector{Vector{I}},I},M} <: LightGraphs.AbstractGraph{I}
     ne::Int
-    fadjlist::Vector{Vector{I}} # `fadjlist[src] => dsts`
-    badjlist::Vector{Vector{I}} # `badjlist[dst] => srcs`
+    fadjlist::F # `fadjlist[src] => dsts`
+    badjlist::B # `badjlist[dst] => srcs` or `ndsts`
     metadata::M
 end
-BipartiteGraph(ne::Integer, fadj::AbstractVector, badj::AbstractVector) = BipartiteGraph(ne, fadj, badj, nothing)
+BipartiteGraph(ne::Integer, fadj::AbstractVector, badj::Union{AbstractVector}=maximum(maximum, fadj); metadata=nothing) = BipartiteGraph(ne, fadj, badj, metadata)
```
Comment (anandijain) [inline src/bipartite_graph.jl:75]: why do you need the `Union{}`? just curious

Comment (YingboMa) [reply src/bipartite_graph.jl:75]: Ah, that's a typo

**Comment by anandijain** [comment]
is it happening? O.O

**Comment by YingboMa** [comment]
The CellMLToolkit failure is caused by https://github.com/SciML/CellMLToolkit.jl/issues/47 . I will merge this PR when other tests pass, because the CellMLToolkit failure is caused by bugs in CellMLToolkit itself.

**Comment by YingboMa** [comment]
The NNPDE test failure looks unrelated.

**Pull request merged**